examples/StandardFirmataChipKIT/StandardFirmataChipKIT.ino

```diff
@@ -0,0 +1,735 @@
+}
+
+// -----------------------------------------------------------------------------
+/* sets the pin mode to the correct state and sets the relevant bits in the
+ * two bit-arrays that track Digital I/O and PWM status
+ */
+void setPinModeCallback(byte pin, int mode)
+{
+  if (pinConfig[pin] == IGNORE)
+    return;
+
+  if (pinConfig[pin] == I2C && isI2CEnabled && mode != I2C) {
+    // disable i2c so pins can be used for other functions
+    // the following if statements should reconfigure the pins properly
+    disableI2CPins();
+  }
+  if (IS_PIN_DIGITAL(pin) && mode != SERVO) {
+    if (servoPinMap[pin] < MAX_SERVOS && servos[servoPinMap[pin]].attached()) {
+      detachServo(pin);
+    }
+  }
+  if (IS_PIN_ANALOG(pin)) {
+    reportAnalogCallback(PIN_TO_ANALOG(pin), mode == ANALOG ? 1 : 0); // turn on/off reporting
+  }
+  if (IS_PIN_DIGITAL(pin)) {
+    if (mode == INPUT) {
+      portConfigInputs[pin / 8] |= (1 << (pin & 7));
+    } else {
+      portConfigInputs[pin / 8] &= ~(1 << (pin & 7));
+    }
+  }
+  pinState[pin] = 0;
+  switch (mode) {
+    case ANALOG:
+      if (IS_PIN_ANALOG(pin)) {
+        if (IS_PIN_DIGITAL(pin)) {
+          pinMode(PIN_TO_DIGITAL(pin), INPUT);    // disable output driver
+          digitalWrite(PIN_TO_DIGITAL(pin), LOW); // disable internal pull-ups
+        }
+        pinConfig[pin] = ANALOG;
+      }
+      break;
+    case INPUT:
+      if (IS_PIN_DIGITAL(pin)) {
+        pinMode(PIN_TO_DIGITAL(pin), INPUT);    // disable output driver
+        digitalWrite(PIN_TO_DIGITAL(pin), LOW); // disable internal pull-ups
+        pinConfig[pin] = INPUT;
+      }
+      break;
+    case OUTPUT:
+      if (IS_PIN_DIGITAL(pin)) {
+        digitalWrite(PIN_TO_DIGITAL(pin), LOW); // disable PWM
+        pinMode(PIN_TO_DIGITAL(pin), OUTPUT);
+        pinConfig[pin] = OUTPUT;
+      }
+      break;
+    case PWM:
+      if (IS_PIN_PWM(pin)) {
+        pinMode(PIN_TO_PWM(pin), OUTPUT);
+        servoWrite(PIN_TO_PWM(pin), 0);
+        pinConfig[pin] = PWM;
+      }
+      break;
+    case SERVO:
+      if (IS_PIN_DIGITAL(pin)) {
+        pinConfig[pin] = SERVO;
+        if (servoPinMap[pin] == 255 || !servos[servoPinMap[pin]].attached()) {
+          // pass -1 for min and max pulse values to use default values set
+          // by Servo library
+          attachServo(pin, -1, -1);
+        }
+      }
+      break;
+    case I2C:
+      if (IS_PIN_I2C(pin)) {
+        // mark the pin as i2c
+        // the user must call I2C_CONFIG to enable I2C for a device
+        pinConfig[pin] = I2C;
+      }
+      break;
+    default:
+      Firmata.sendString("Unknown pin mode"); // TODO: put error msgs in EEPROM
+  }
+  // TODO: save status to EEPROM here, if changed
+}
+
+void analogWriteCallback(byte pin, int value)
+{
+  if (pin < TOTAL_PINS) {
+    switch (pinConfig[pin]) {
+      case SERVO:
+        if (IS_PIN_DIGITAL(pin))
+          servos[servoPinMap[pin]].write(value);
```

Comment:
Hmm. I guess I'm not quite following. Any pin that can be a digital I/O on
PIC32 can be PWM or Servo outputs as well (using SoftPWMServo library) so
there shouldn't be any reason for the client libraries to worry about what
pins they use for PWM or Servo outputs. Or am I missing the point?

On Fri, Aug 7, 2015 at 10:02 PM, Jeff Hoefs <notifications@github.com>
wrote:

> In examples/StandardFirmataChipKIT/StandardFirmataChipKIT.ino
> <https://github.com/firmata/arduino/pull/210#discussion_r36574396>:
>
> > +        pinConfig[pin] = I2C;
> > +      }
> > +      break;
> > +    default:
> > +      Firmata.sendString("Unknown pin mode"); // TODO: put error msgs in EEPROM
> > +  }
> > +  // TODO: save status to EEPROM here, if changed
> > +}
> > +
> > +void analogWriteCallback(byte pin, int value)
> > +{
> > +  if (pin < TOTAL_PINS) {
> > +    switch (pinConfig[pin]) {
> > +      case SERVO:
> > +        if (IS_PIN_DIGITAL(pin))
> > +          servos[servoPinMap[pin]].write(value);
>
> This will break if anyone tries to write to a servo using any existing
> Firmata client library that supports servos. You will need to move
> servoWrite here and handle and PWM conversion as necessary or it will not
> be compatible with existing Firmata client libs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firmata/arduino/pull/210/files#r36574396>.
>